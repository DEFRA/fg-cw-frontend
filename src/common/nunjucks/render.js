import camelCase from "lodash/camelCase.js";
import { environment } from "./nunjucks.js";

/**
 * @param {string} name
 * @param {object} params
 * @param {string} [block]
 */
export const render = (name, params, block) => {
  const macro = {
    path: `${name}/macro.njk`,
    name: camelCase(name.split("-")),
    params: JSON.stringify(params, null, 2),
  };

  const callComponent = block
    ? `{%- call ${macro.name}(${macro.params}) -%}${block}{%- endcall -%}`
    : `{{- ${macro.name}(${macro.params}) -}}`;

  return environment.renderString(
    `{%- from "${macro.path}" import ${macro.name} -%}
    ${callComponent}
    `,
    {},
  );
};
