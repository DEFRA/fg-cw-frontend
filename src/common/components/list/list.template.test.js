import { describe, expect, test } from "vitest";
import { render } from "../../nunjucks/render.js";

describe("list", () => {
  test("renders list with fields and payload", () => {
    const fields = [
      {
        ref: "$.payload.answers.totalPigs",
        type: "number",
        label: "Total Pigs",
      },
      {
        ref: "$.payload.answers.isPigFarmer",
        type: "boolean",
        label: "Are you a pig farmer?",
      },
    ];

    const payload = {
      answers: {
        totalPigs: 10,
        isPigFarmer: true,
      },
    };

    const options = {
      fields,
      payload,
    };

    const component = render("list", options);
    expect(component).toMatchSnapshot();
  });

  test("renders nothing when no fields provided", () => {
    const component = render("list", {});
    expect(component.trim()).toBe("");
  });

  test("renders nothing when no payload provided", () => {
    const fields = [
      {
        ref: "$.payload.answers.totalPigs",
        type: "number",
        label: "Total Pigs",
      },
    ];

    const component = render("list", { fields });
    expect(component.trim()).toBe("");
  });
});
