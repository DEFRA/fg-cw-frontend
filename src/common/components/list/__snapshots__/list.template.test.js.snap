// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`List component > renders list with fields and payload 1`] = `undefined`;

exports[`list > renders list with fields and payload 1`] = `
"
    
    
    
    
    
    
    
    

  <table class="govuk-table">
  <tbody class="govuk-table__body">
    <tr class="govuk-table__row">
      <th scope="row" class="govuk-table__header govuk-!-width-one-third">Total Pigs</th>
      <td class="govuk-table__cell">10</td>
    </tr>
    <tr class="govuk-table__row">
      <th scope="row" class="govuk-table__header govuk-!-width-one-third">Are you a pig farmer?</th>
      <td class="govuk-table__cell">Yes</td>
    </tr>
  </tbody>
</table>

 "
`;
